File: makeConnectedNetwork.java
import java.util.*;
// Approach 1- DFS (Recursive)
class Solution{ // TC O(n+e) SC O(n+e)
    public void dfs(int node, Map<Integer, List<Integer>> adj, boolean[] visit) {
        visit[node] = true;
        if (!adj.containsKey(node)) {
            return;
        }
        for (int neighbor : adj.get(node)) {
            if (!visit[neighbor]) {
                visit[neighbor] = true;
                dfs(neighbor, adj, visit);
            }
        }
    }

    public int makeConnected(int n, int[][] connections) {
        if (connections.length < n - 1) {
            return -1;
        }

        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int[] connection : connections) {
            adj.computeIfAbsent(connection[0], k -> new ArrayList<Integer>()).add(connection[1]);
            adj.computeIfAbsent(connection[1], k -> new ArrayList<Integer>()).add(connection[0]);
        }

        int numberOfConnectedComponents = 0;
        boolean[] visit = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visit[i]) {
                numberOfConnectedComponents++;
                dfs(i, adj, visit);
            }
        }

        return numberOfConnectedComponents - 1;
    }
}

// Approach 2- BFS 
class Solution1{ // TC O(n+e) SC O(n+e)
    public void bfs(int node, Map<Integer, List<Integer>> adj, boolean[] visit) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(node);
        visit[node] = true;

        while (!q.isEmpty()) {
            node = q.poll();
            if (!adj.containsKey(node)) {
                continue;
            }
            for (int neighbor : adj.get(node)) {
                if (!visit[neighbor]) {
                    visit[neighbor] = true;
                    q.offer(neighbor);
                }
            }
        }
    }

    public int makeConnected(int n, int[][] connections) {
        if (connections.length < n - 1) {
            return -1;
        }

        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int[] connection : connections) {
            adj.computeIfAbsent(connection[0], k -> new ArrayList<Integer>()).add(connection[1]);
            adj.computeIfAbsent(connection[1], k -> new ArrayList<Integer>()).add(connection[0]);
        }

        int numberOfConnectedComponents = 0;
        boolean[] visit = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visit[i]) {
                numberOfConnectedComponents++;
                bfs(i, adj, visit);
            }
        }

        return numberOfConnectedComponents - 1;
    }
}

// Approach 3- Union-Find (Disjoint sets)
class Solution2{ // TC O(n+e) SC O(n)
    public int find(int vertex, int[] parent){
        if(parent[vertex]!=vertex){
            parent[vertex]=find(parent[vertex], parent);
        }
        return parent[vertex];
    }
    
    
    public void union(int v1, int v2, int[] parent, int[] size){
        v1=find(v1, parent);
        v2=find(v2, parent);
        if(v1!=v2){
            if(size[v1]>size[v2]){
                size[v1]+=size[v2];
                parent[v2]=v1;
            } else {
                size[v2]+=size[v1];
                parent[v1]=v2;
            }
        }
        return;
    }
    
    public int makeConnectedNetwork(int n, int[][] connections) {
        
        if(connections.length < n-1) {
            return -1;
        }
        
        int[] parent=new int[n];
        int[] size=new int[n];
        
        for(int i=0; i<n; i++){
            parent[i]=i;
            size[i]=1;
        }
        
        int answer=n;
        for(int[] connection: connections){
            if(find(connection[0], parent)!=find(connection[1], parent)){
                answer--;
                union(connection[0], connection[1], parent, size);
            }
        }
        return answer-1;
    }
}

class makeConnectedNetwork{
    public static void main(String[] args) {
        int[][] connections={{0,1},{0,2},{1,2}};
        int n=4;
        Solution s=new Solution();
        System.out.println(s.makeConnectedNetwork(n, connections));
    }
}

File: makeConnectedNetwork.py
class Solution:
# Approach 1- DFS
    def makeConnectedNetwork(self, n, connections): # TC O(n+e) // SC O(n+e)
            def dfs(node):
                visit[node]=True
                if node not in adj:
                    return
                
                for neighbor in adj[node]:
                    if not visit[neighbor]:
                        visit[neighbor]=True
                        dfs(neighbor)
    
        
        if len(connections)<n-1: return -1
        
        adj = {}
        for connection in connections:
            if connection[0] not in adj:
                adj[connection[0]] = []
            if connection[1] not in adj:
                adj[connection[1]] = []
            adj[connection[0]].append(connection[1])
            adj[connection[1]].append(connection[0])
        
        visit=[False]*(len(connections)+1)
        answer=0
        for i in range(n):
            if not visit[i]:
                answer+=1
                dfs(i)
        
        return answer-1

# Approach 2- BFS
    def makeConnectedNetwork(self, n, connections): # TC O(n+e) // SC O(n+e)
            def bfs(node):
                q=deque()
                q.append(node)
                visit[node]=True
                while q:
                    node=q.popleft()
                    if node not in adj: continue 
                    
                    for neighbor in adj[node]:
                        if not visit[neighbor]:
                            visit[neighbor]=True
                            q.append(neighbor)
                
                return

        if len(connections)<n-1: return -1
        
        adj = {}
        for connection in connections:
            if connection[0] not in adj:
                adj[connection[0]] = []
            if connection[1] not in adj:
                adj[connection[1]] = []
            adj[connection[0]].append(connection[1])
            adj[connection[1]].append(connection[0])
        
        visit=[False]*(len(connections)+1)
        answer=0
        for i in range(n):
            if not visit[i]:
                answer+=1
                bfs(i)
        
        return answer-1


# Approach 3- Union-Find (Disjoint sets)    
    def makeConnectedNetwork(self, n, connections): # TC O(n+e) // SC O(n)
        
        if len(connections) <n-1: return -1
        
        def find(vertex):
            if parent[vertex]!=vertex:
                parent[vertex]=find(parent[vertex])
            
            return parent[vertex] 
        
        def union(v1, v2):
            v1=find(v1)
            v2=find(v2)
            
            if v1!=v2:
                if size[v1]>size[v2]:
                    size[v1]+=size[v2]
                    parent[v2]=v1
                else:
                    size[v2]+=size[v1]
                    parent[v1]=v2
            
            return 
        
        parent=[i for i in range(n)]
        size=[1 for i in range(n)]
        answer=n
        
        for conn in connections:
            if find(conn[0])!=find(conn[1]):
                answer-=1
                union(conn[0],conn[1])
        
        return answer-1


s=Solution()
print(s.makeConnectedNetwork(4,[[0,1],[0,2],[1,2]]))